quinn/src/connection.rs

```diff
@@ -287,25 +287,24 @@ where
         let span = info_span!("drive", id = conn.handle.0);
         let _guard = span.enter();
 
-        loop {
-            let mut keep_going = false;
-            if let Err(e) = conn.process_conn_events(cx) {
-                conn.terminate(e);
-                return Poll::Ready(());
-            }
-            conn.drive_transmit();
-            // If a timer expires, there might be more to transmit. When we transmit something, we
-            // might need to reset a timer. Hence, we must loop until neither happens.
-            keep_going |= conn.drive_timer(cx);
-            conn.forward_endpoint_events();
-            conn.forward_app_events();
-            if !keep_going || conn.inner.is_drained() {
-                break;
-            }
+        if let Err(e) = conn.process_conn_events(cx) {
+            conn.terminate(e);
+            return Poll::Ready(());
         }
+        let mut keep_going = conn.drive_transmit();
+        // If a timer expires, there might be more to transmit. When we transmit something, we
+        // might need to reset a timer. Hence, we must loop until neither happens.
+        keep_going |= conn.drive_timer(cx);
+        conn.forward_endpoint_events();
+        conn.forward_app_events();
 
         if !conn.inner.is_drained() {
-            conn.driver = Some(cx.waker().clone());
+            if keep_going {
+                // If the connection hasn't processed all tasks, schedule it again
+                cx.waker().wake_by_ref();
```

Comment:
This means when we get a timer wakeup (e.g. due to keep-alives) we'll have to go through two wakeup cycles to transmit a packet in response. Is that desired? Admittedly that's not the hot path...

Reply:
Yes, I wondered about that too. But given it's not a regular event I didn't feel it was worth keeping the loop just for that case. I can add it back if others feel different about it. 

However I also thought it would only matter if a timer expired immediately or between the start of `drive` and the timer call. Now I see that this is not true, and that all timer expirations would require 2 iterations since timer expiration is not checked unless the tokio timer asks us to do so. That makes it kind of more interesting to add the loop back (or to move the timer check between processing incoming events and sending outgoing events?)

Reply:
> or to move the timer check between processing incoming events and sending outgoing events

We update the loss detection and pacing timers when sending a packet, so if we move the timers before `poll_transmit` we'll need far more double-loops than the current configuration.

Most timer expirations are rare, but the pacing timer might be firing ~continuously when the network is saturated. Might be worth gathering some metrics to inform this. Though, empirically the impact isn't that large, so we don't have to make a huge deal out of it.